cpp/src/arrow/table_test.cc

```diff
@@ -406,6 +407,171 @@ TEST_F(TestTable, ConcatenateTables) {
   ASSERT_RAISES(Invalid, ConcatenateTables({t1, t3}, &result));
 }
 
+class TestPromoteTableToSchema : public TestTable {
+ protected:
+  std::shared_ptr<Table> MakeTableWithOneNullColumn(const std::string& column_name,
+                                                    const int length) {
+    return Table::Make(schema({field(column_name, null())}),
+                       {MakeRandomArray<NullArray>(length)});
+  }
+
+  std::shared_ptr<Table> MakeTableWithOneNullFilledColumn(
+      const std::string& column_name, const std::shared_ptr<DataType>& data_type,
+      const int length) {
+    std::shared_ptr<Array> array_of_nulls;
+    DCHECK_OK(MakeArrayOfNull(data_type, length, &array_of_nulls));
```

Comment:
nit: MakeArrayOfNull supports Type::NULL so `MakeTableWithNullColumn("name", length)` could be replaced with `MakeTableWithOneNullFilledColumn("name", null(), length)`